pkg/allocation/libs/converter.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2019 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package converters includes API concersions between GameServerAllocation API and the Allocation proto APIs
+package converters
+
+import (
+	pb "agones.dev/agones/pkg/allocation/go/v1alpha1"
+	"agones.dev/agones/pkg/apis"
+	agonesv1 "agones.dev/agones/pkg/apis/agones/v1"
+	allocationv1 "agones.dev/agones/pkg/apis/allocation/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ConvertAllocationRequestV1Alpha1ToGSAV1 converts AllocationRequest to GameServerAllocation V1 (GSA)
```

Comment:
Is GSA v1 the "canonical" form? When we introduce v1beta1 or v1 for the allocation api how much more conversion code will we need to add?

Reply:
If we diverge to multiple versions of GSA and allocation.proto, there may be multiple conversions. However, ideally the conversions will always happen to the latest GSA and we handle compatibility at this layer. In that case there should be an internal version of GSA.